src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java

```diff
@@ -640,6 +640,9 @@ private boolean isNullnessSubtype(AnnotatedTypeMirror subtype, AnnotatedTypeMirr
           && isNullnessSubtype(subtype, ((AnnotatedTypeVariable) supertype).getLowerBound())) {
         return true;
       }
+      if (!isLeastConvenientWorld && subtype.hasEffectiveAnnotation(nullnessOperatorUnspecified)) {
```

Comment:
tl;dr See "**My implementation suggestion is this**" below. It makes the same change but restricted to null types. We can revisit when we discuss JSpecify's odd substitution behavior again someday.

Dumping details here:

```
/tmp/tmp.zQJso3R0U8/jspecify-reference-checker/tests/regression/Issue172.java:33:Error: (type.argument.type.incompatible) incompatible type argument for type parameter E of Issue172.
void foo(Issue172<Object> p) {}
                  ^
found   : Object
required: [extends Object* super null*]
```

Seemingly `null*` is not a subtype of `Object`, even in lenient mode. I am now trying to decide whether that makes sense or not....

I think we've seen `null*` [rarely](https://jspecify.dev/docs/spec/#null-types). So we might just not handle it well.

Continuing to think this through in too much detail for a PR comment:

```java
@NullUnmarked
<T> T unspecify(@Nullable T t);

s = unspecify(null);
```

At that point, I'd expect `s` to have type `null*`. Or really, it would have some type that depends on the declaration of `s`. So if it were a field `String s` in `@NullUnmarked` code, then we'd see `T` inferred to be either `String` or `String*`, and the result of the call to `unspecify` would be of type `String*`.

I wonder if there is actually _any_ scenario in which we'd see `null*`? I feel like I once long ago ran all our tests to look for `null*` in the error messages, and I don't remember whether I found any. Hmm, but I'm pretty sure that I [see null types sometimes in the javac APIs](https://github.com/google/error-prone/blob/5a7b8d9b41a19aaf6cc917bc295ab5201cc2f328/core/src/main/java/com/google/errorprone/bugpatterns/nullness/NullnessUtils.java#L546), so it probably does come up, at least for types of _expressions_ (though probably still not variables, etc.).

It's weird... if there is no type information available for some call to `unspecify`, then I think(?) that method can't _really_ fulfill its contract when someone calls `unspecify(null)`—except by returning `null`. So there is some sense in not going to any effort to make a `null*` type behave any differently than `null?`.

But `null*` probably can already behave differently than `null?` in at least some cases, since the latter is always null-inclusive. And clearly `null*` means neither "the bottom type" nor "null" but some sort of ambiguous state, which might also arise in other ways besides what we've seen here. So I guess `null*` probably should be a subtype of `Object` in lenient mode?

Or maybe it still shouldn't? If I tweak my `unspecify` example above to declare its type parameter as `<T extends @Nullable Object>`, then, if we infer a null type, then that might make more sense to treat the return type like "`T*`, where `T` is still potentially a nullable type?" But now we're back into the weird JSpecify substitution behavior (under which `T*` can "become nullable" after substitution), so I'm going to get off that train of thought now.

All that I think really need to do here is to get things working, _ideally_ with some general understanding of what we're doing: The checker is already full of as many unprincipled hacks as I could cram in, and I'm OK with that if I feel like I'm learning something. So let's make these tests pass.

**My implementation suggestion is this**: Change `IS_DECLARED_OR_ARRAY` to `IS_DECLARED_OR_ARRAY_OR_NULL` (and `IsDeclaredOrArray` to `IsDeclaredOrArrayOrNull`). The "actual" code change will be to add ` || t.getKind() == NULL` to the end of the conditional. Then the check in `isNullExclusiveUnderEveryParameterization` will do the "right" thing, and we won't need the change here.

The advantage to that, aside from requiring slightly less code, is that it preserves the (again, questionable) JSpecify substitution behavior. For example, it lets the following code remain an error:

```java
interface Foo<E extends @Nullable Object> {
  E e();
}

@NullMarked
class Bar<E extends @Nullable Object> {
  Object o(Foo<E> foo) {
    return foo.e(); // `E*` is not compatible with `Object` even in lenient mode
  }
}
```

Hopefully that will give you fewer changes in samples behavior to sort through.

Reply:
I have successfully talked myself into liking my implementation suggestion's special case a lot more than I did before.

My claim above was that the checker might analyze an `unspecify` method and conclude that the method could transform `null?` into `null*`—and that that seemed like a promise that it could never fulfill.

But in hindsight, there's an obvious way to fulfill that promise: `unspecify` just needs to behave like `requireNonNull`! Such a method is transforming `null?` into `null` (to put it in the checker's current format), the bottom type. (I wouldn't be shocked if `requireNonNull(null)` is considered by our checker to have the type `null` as things stand today, at least in some contexts. Not that I'd expect to see that particular code—except in some tests of `requireNonNull` or tools' handling of it!)

And I think I should have reached a similar conclusion by focusing more on pure theory: If I focus on `null` as the bottom type, which has no values, then clearly it is a subtype of everything. And if `null*` might "really be" `null`, then it, too, might be a subtype of everything.

Basically, I had been viewing the existing handling of declared types and array types as the "special case," when in fact they're no different from the null type. Arguably the true "special cases" are those in which we need to "look through" types to their components, as with intersection types, type variables, and wildcards.

(This might actually call for a spec update....)